### PR TITLE
Add `Trainer.init_module` and `LightningModule.configure_model`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Port of https://github.com/Lightning-AI/lightning/pull/17462 for the Trainer

Key implementation differences:
- No `init_context` support for `DeepSpeedPrecisionPlugin` because it doesn't support true precision (unlike Fabric)
- `Trainer` only exposes `init_module` which functionally matches `fabric.init_tensor` because otherwise it would require that processes have been launched when it is called, but that won't happen until `trainer.fit()`.
- [NEW HOOK] Instead of `Trainer.init_module`, users are suggested `LightningModule.configure_model` with sharded strategies 
- `LightningModule.configure_sharded_model` is deprecated in favor of `LightningModule.configure_model`

Generated docs: https://pytorch-lightning--18004.org.readthedocs.build/

Enables https://github.com/Lightning-AI/lit-gpt/pull/228

cc @borda @awaelchli @carmocca @justusschock

### CODE REVIEW

File: src/lightning/pytorch/strategies/fsdp.py [hunk 1]
```diff
@@ -262,13 +261,7 @@ def setup(self, trainer: "pl.Trainer") -> None:
         assert self.lightning_module is not None
         self.lightning_module._device = self.root_device
 
-        if is_overridden("configure_sharded_model", self.lightning_module):
```
Comment (carmocca) [inline src/lightning/pytorch/strategies/fsdp.py:265]: Now that there are uses for `configure_sharded_model` that aren't wrapping, we shouldn't skip wrapping if the hook is defined.

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 2]
```diff
@@ -67,6 +67,39 @@ For training on unreliable mixed GPUs across the internet check out the :doc:`Hi
 ----
 
 
+************************
+Efficient initialization
```
Comment (carmocca) [inline docs/source-pytorch/advanced/model_parallel.rst:71]: Is there a better doc to put this section? Open to suggestions

File: src/lightning/fabric/strategies/fsdp.py [hunk 3]
```diff
@@ -222,24 +222,27 @@ def setup_module(self, module: Module) -> "FullyShardedDataParallel":
...
-            module=module,
-            cpu_offload=self.cpu_offload,
-            mixed_precision=self.mixed_precision_config,
-            device_id=self.root_device.index,
-            **self._fsdp_kwargs,
-        )
+        if any(isinstance(mod, FullyShardedDataParallel) for mod in module.modules()):
+            # the user wrapped at least one layer in `configure_sharded_model` already
+            if "auto_wrap_policy" in self._fsdp_kwargs:
+                rank_zero_warn(
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:228]: Newly added warning

File: src/lightning/fabric/strategies/fsdp.py [hunk 4]
```diff
@@ -222,24 +222,27 @@ def setup_module(self, module: Module) -> "FullyShardedDataParallel":
...
-            # If model is already wrapped, we need to avoid sending the `auto_wrap_policy`
-            del self._fsdp_kwargs["auto_wrap_policy"]
-        wrapped_module = FullyShardedDataParallel(
-            module=module,
-            cpu_offload=self.cpu_offload,
-            mixed_precision=self.mixed_precision_config,
-            device_id=self.root_device.index,
-            **self._fsdp_kwargs,
-        )
+        if any(isinstance(mod, FullyShardedDataParallel) for mod in module.modules()):
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:225]: We don't wrap the root if any module is already wrapped. I tested it and it seems to work.

File: src/lightning/fabric/strategies/fsdp.py [hunk 5]
```diff
@@ -594,13 +597,26 @@ def _set_world_ranks(self) -> None:
...
+        CheckpointWrapper,
     )
 
+    if any(isinstance(mod, CheckpointWrapper) for mod in module.modules()):
+        if layers:
+            rank_zero_warn(
+                f"FSDP checkpointing for the layers {layers} is configured, but the model already contains checkpointed"
+                " layers. Checkpointing will be ignored."
+            )
+        # the module is already wrapped with activation checkpointing, avoid wrapping twice
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:614]: I tested it, it's possible to wrap twice

File: tests/tests_pytorch/strategies/test_common.py [hunk 6]
```diff
@@ -48,21 +47,3 @@ def test_evaluate(tmpdir, trainer_kwargs):
     # make sure weights didn't change
     new_weights = model.layer_0.weight.clone().detach().cpu()
     torch.testing.assert_close(old_weights, new_weights)
-
-
-def test_model_parallel_setup_called(tmpdir):
```
Comment (carmocca) [inline tests/tests_pytorch/strategies/test_common.py:53]: We already have tests that check that the hook is called

File: docs/source-pytorch/advanced/model_parallel.rst (hunk 2 above)
Comment (awaelchli) [reply docs/source-pytorch/advanced/model_parallel.rst:71]: I have this idea:

1. Make an entire standalone page alone just for this (what you described here). 
2. In addition, mention only the relevant part about "configure_model" here in the model parallel docs. 
3. Add relevant info to the mixed precision docs page
4. Add relevant info in the model checkpoint loading docs

What do you think?

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 7]
```diff
@@ -67,6 +67,54 @@ For training on unreliable mixed GPUs across the internet check out the :doc:`Hi
...
+
+.. code-block:: python
+
+    class MyModel(LightningModule):
+        def __init__(self):
+            super().__init__()
+            # don't instantiate layers here
+            # move the creation of layers to `configure_model`
+
+        def configure_model(self):
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/model_parallel.rst:111]: This name makes sense considering that the previous one was `configure_sharded_model`. I would still suggest a different name because `configure_x` already has a different meaning in Lightning, i.e., the configure_optimizers hook creates the optimizer and returns it. Furthermore, "configuration" refers to something that already exists and gets configured. For these reasons, I'm wondering should we choose a different name? 

The code that goes into this hook is basically what normally is the layer creation in the standard `__init__`. Here are a couple proposals we could think about?

- create_layers
- init_layers (might be confused with weight initializtion???)
- init_model

No strong opinion.

File: src/lightning/fabric/CHANGELOG.md [hunk 8]
```diff
@@ -41,12 +41,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   * Handles initialization for FSDP models before wrapping and the Zero stage 3 initialization for DeepSpeed before sharding
 
 
-- Added supports for empty weight initialization with `Fabric.init_module(empty_init=True)` for efficient sharding and checkpoint loading ([#17627](https://github.com/Lightning-AI/lightning/pull/17627))
```
Comment (awaelchli) [inline src/lightning/fabric/CHANGELOG.md:44]: Why was this deleted? Please keep it. Of course not all strategies support it, but it is useful for stage 3 training for sure.

File: src/lightning/pytorch/trainer/call.py [hunk 9]
```diff
@@ -93,9 +94,17 @@ def _call_setup_hook(trainer: "pl.Trainer") -> None:
...
-    with trainer.strategy.model_sharded_context():
-        _call_lightning_module_hook(trainer, "configure_sharded_model")
+def _call_configure_model(trainer: "pl.Trainer") -> None:
+    # legacy hook
+    if is_overridden("configure_sharded_model", trainer.lightning_module):
+        with trainer.strategy.model_sharded_context():
+            _call_lightning_module_hook(trainer, "configure_sharded_model")
+
+    # we don't normally check for this before calling the hook. it is done here to avoid instantiating the context
+    # managers
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/call.py:104]: But the context manager doesn't do anything if the hook doesn't end up being called?

File: src/lightning/pytorch/trainer/trainer.py [hunk 10]
```diff
@@ -1068,6 +1071,41 @@ def __setup_profiler(self) -> None:
...
+                Set this to ``True`` if you are loading a checkpoint into a large model. Requires `torch >= 1.13`.
+        """
+        if not _TORCH_GREATER_EQUAL_2_0 and self.strategy.root_device.type != "cpu":
+            rank_zero_warn(
+                "`Trainer.init_module()` can't place tensors on the device directly"
+                " with PyTorch < 2.0. Parameters will remain on CPU until the trainer starts."
+                " Upgrade to PyTorch >= 2.0 to fully utilize this feature.",
+                category=PossibleUserWarning,
+            )
+        if is_overridden("model_sharded_context", self.strategy, parent=Strategy):
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/trainer.py:1095]: I think this is too often a false positive. While DeepSpeed overrides this method, it really only applies the context manager for stage 3. So for stage 1 and 2 this warning would be a false positive always. Same for FSDP with non-standard configuration. You'd need to reach into the strategies and check the configuration.
For simplicity, I recommend not doing this warning.

File: tests/tests_pytorch/strategies/test_fsdp.py [hunk 11]
```diff
@@ -627,3 +626,38 @@ def test_fsdp_strategy_load_optimizer_states(tmpdir, wrap_min_params):
...
+        devices=2,
+        strategy=FSDPStrategy(auto_wrap_policy=always_wrap_policy),
+        precision=precision,
+        fast_dev_run=1,
+    )
+
+    class MyModel(BoringModel):
+        def configure_model(self):
+            self.layer = torch.nn.Linear(32, 2)
+            # The model is on the CPU until `.setup()``
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_fsdp.py:651]:
```suggestion
            # The model is on the CPU until after `.setup()``
```
nit :)

**Review by awaelchli — COMMENTED** [review]
Looks great. Minor questions and comments only

File: src/lightning/fabric/CHANGELOG.md (hunk 8 above)
Comment (carmocca) [reply src/lightning/fabric/CHANGELOG.md:44]: I thought it was wrong: EmptyInit is not used with DS zero stage 3. It's managed by DeepSpeed internally. With this I mean that the behaviour was is unchanged after the addition of `empty_init`. In the code we have

`super().module_init_context(empty_init=empty_init) if not self.zero_stage_3 else nullcontext()`

File: src/lightning/pytorch/trainer/call.py (hunk 9 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/call.py:104]: Even though what you said is true today, I'd rather avoid this unnecessary step so that we don't risk silent changes if there's a side effect inside FSDP or DeepSpeed in the future. We have tests that the context manager is only called once.

File: src/lightning/pytorch/trainer/trainer.py (hunk 10 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/trainer.py:1095]: When users are using `trainer.init_module` with non-sharded and switch to FSDP, how do they find out that now they need to change their code and instantiate a hook? Just by knowing this from the docs? If they don't know this detail FSDP will silently not work and that is bad

File: docs/source-pytorch/advanced/model_parallel.rst (hunk 7 above)
Comment (carmocca) [reply docs/source-pytorch/advanced/model_parallel.rst:111]: I assumed that if `configure_sharded_model` was originally approved, then removing `sharded` would be too :wink:. I'll go with whatever is chosen. My personal thought is that using "mod**el**" is not consistent with "init_mod**ule**"

File: docs/source-pytorch/advanced/model_parallel.rst (hunk 2 above)
Comment (carmocca) [reply docs/source-pytorch/advanced/model_parallel.rst:71]: Do (2), (3), and (4) duplicate part of the information that is in (1)?

If we are thinking of refactoring the overall structure, I would prefer to leave this for a follow-up. I can open an issue about it

File: src/lightning/pytorch/trainer/trainer.py (hunk 10 above)
Comment (awaelchli) [reply src/lightning/pytorch/trainer/trainer.py:1095]: So how would you address my comment then?

File: src/lightning/fabric/CHANGELOG.md (hunk 8 above)
Comment (awaelchli) [reply src/lightning/fabric/CHANGELOG.md:44]: **our** empty init is not used with stage 3 because it is not needed when the context manager for stage 3 uses the fake tensors anyway. That's why we are setting empty_init=None as default for DeepSpeed and validate the use depending on the configured stage. The empty_init argument in the `init_module` method is not to be directly associated with our _EmptyInit implementation. 

The whole point of all this work was to do this, and now deleting it from the changelog is strange IMO.

File: docs/source-pytorch/advanced/model_parallel.rst (hunk 2 above)
Comment (awaelchli) [reply docs/source-pytorch/advanced/model_parallel.rst:71]: You directly asked for suggestions, this was my suggestion. 

The feature is useful for a variety of things, putting it in a specific model parallel training doc page is not ideal IMO. Putting it in relevant places or having a separate page is not so far-fetched of an idea, considering the existing documentation structure we have.

File: src/lightning/pytorch/trainer/trainer.py (hunk 10 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/trainer.py:1095]: So far the options are:
(1): Have no warning
(2): Have the warning, accept false positives
(3): Have the warning, add logic to reduce false positives
(4): don't add this method

All of them have their flaws. I would go with (2) which is what's currently implemented

File: src/lightning/fabric/CHANGELOG.md (hunk 8 above)
Comment (carmocca) [reply src/lightning/fabric/CHANGELOG.md:44]: What I'm saying is that changing `empty_init` does not change whether sharding is more or less efficient today:
- FSDP doesn't support meta device instantiation at the moment
- DeepSpeed Zero stage 3 ignores the `empty_init` argument. It uses fake tensors anyways

In the future, this might not be the case anymore and then it will help with efficient sharding. But today that snippet in the CHANGELOG is misleading

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
